1.MLP_Binary_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,881 @@
+        The number of epochs to train the model for
+    train_loader : torch.utils.data.DataLoader
+        DataLoader for train data integration to pytorch
+    valid_loader : torch.utils.data.DataLoader
+        DataLoader for train data integration to pytorch
+    in_features : int
+        the number of in features for the initial network layer
+    out_features : int
+        the number of out features for the final network layer
+    parameter_dict : dict
+        dictionary of optimized model hyperparameters
+    params : object
+        Dataclass containing constants and parameter spaces
+
+    Returns
+    -------
+    tuple[float, float, float, float, int, object]
+        train_loss: float
+        train_acc: float
+        valid_loss: float
+        valid_acc: float
+        epochs_ran: int
+        model: torch.nn.Sequential
+
+    """
+
+    model = optimized_model_create(in_features, out_features, parameter_dict)
+    model = model.to(params.DEVICE)
+    # criterion is the method in which we measure our loss
+    # isn't defined as loss as it doesn't represent the loss value but the method
+    criterion = nn.BCEWithLogitsLoss()
+    optim_method = parameter_dict["optimizer"].strip("'")
+    print(optim_method)
+
+    optimizer = f'optim.{optim_method}(model.parameters(), lr={parameter_dict["learning_rate"]})'
+
+    # optimizer = (
+    # f'optim.{optim_method}(model.parameters(), lr=parameter_dict["learning_rate"])'
+    # )
+    # print(optimizer)
+    optimizer = eval(optimizer)
+    # optimizer = ast.literal_eval(optimizer)
+    print(optimizer)
+
+    early_stopping_patience = 15
+    early_stopping_counter = 0
+
+    train_acc = []
+    train_loss = []
+
+    valid_acc = []
+    valid_loss = []
+
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    valid_loss_min = np.inf
+
+    epochs_ran = []
+
+    for epoch in range(EPOCHS):
+
+        epochs_ran.append(epoch + 1)
+
+        (
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+
+        if np.mean(valid_loss) <= valid_loss_min:
+            torch.save(model.state_dict(), f"./state_dict.pt")
+            print(
+                f"Epoch {epoch + 0:01}: Validation loss decreased ({valid_loss_min:.6f} --> {np.mean(valid_loss):.6f}).  Saving model ..."
+            )
+            valid_loss_min = np.mean(valid_loss)
+            early_stopping_counter = 0  # reset counter if validation loss decreases
```

Comment:
we had previously discussed using a pytorch implementation for early stopping. Just noting again here, this is sufficient for now, just maybe work for the future

Reply:
adding to issues